### PR TITLE
VZ-9336

### PR DESCRIPTION
**Pull request by George-Aeillo**
In this PR, I add dynamic certificate diagnosability to the VZ-analyze tool. I have edited the bug-report code to generate a certificates.json file when a namespace that is captured by the bug-report tool has at least one cert-manager certificate object in its namespace. I then altered the VZ-analysis tool to look at the certificate.json file to report if certificates are expired, the certificates are valid, but the VPO is hanging, and if the certificates are not valid for unspecified reasons. This also outputs of the bug-report to scrub hostnames from all resources.

### CODE REVIEW

File: tools/vz/pkg/analysis/internal/util/cluster/certificates.go [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+	"github.com/verrazzano/verrazzano/tools/vz/pkg/analysis/internal/util/files"
+	"github.com/verrazzano/verrazzano/tools/vz/pkg/analysis/internal/util/report"
+	"go.uber.org/zap"
+)
+
+// This is the initial entry function for certificate related issues.
+// It first determines the status of the VPO, then checks if there are any certificates in the namespaces.
+// It then analyzes those certificates to determine expiration or other issues and then contributes the respective issues to the Issue Reporter.
+// The three issues that it is currently reporting on are the VPO hanging due to a long time to issues validate certificates, expired certificates, and when the certificate is not in a ready status.
+func AnalyzeCertificateRelatedIsssues(log *zap.SugaredLogger, clusterRoot string) (err error) {
```
Comment (davisgomes) [inline tools/vz/pkg/analysis/internal/util/cluster/certificates.go:25]: One extra "s" in this function name

File: tools/vz/pkg/analysis/internal/util/cluster/certificates.go [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+						reportCertificateIssue(log, clusterRoot, certificate, &issueReporter, certificateFile, true, false)
+					}
+				}
+			} else if certificate.Status.NotAfter.Unix() < time.Now().Unix() {
+				reportCertificateIssue(log, clusterRoot, certificate, &issueReporter, certificateFile, false, true)
+			} else {
+				reportCertificateIssue(log, clusterRoot, certificate, &issueReporter, certificateFile, false, false)
+			}
+		}
+	}
```
Comment (davisgomes) [inline tools/vz/pkg/analysis/internal/util/cluster/certificates.go:62]: Typically, it is considered an anti-pattern in Go to nest this deeply. There are a few prescribed methods for resolving nested for loops and if statements:

1. Extrapolate out nested loops to their own function. It is worthwhile in this case to consider making a function that accepts a list of certificates for a namespace and produces a certificate issue report.
2. Reverse if statements. Inverting the clauses for the if statements allows you to reduce nesting. Here is a short example:
```
if statement1 {
  if statement2 {
    // do something
  }
}
```
versus:
```
if !statement1 {
  return
}
if !statement2 {
  return
}
// do something
```
3. reorganize if statements that can be combined. Another example:
```
if statement1 {
  if statement2 {
    // do something
  }
}
```
to
```
if statement1 && statement2 {
    // do something
}
```

Comment (davisgomes) [reply tools/vz/pkg/analysis/internal/util/cluster/certificates.go:62]: Here's a more comprehensive list of techniques: https://en.wikibooks.org/wiki/Computer_Programming/Coding_Style/Minimize_nesting

File: tools/vz/pkg/analysis/internal/util/cluster/certificates.go [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+		certificateListForNamespace, err := getCertificateList(log, certificateFile)
+		if err != nil {
+			return err
+		}
+		if certificateListForNamespace == nil {
+			continue
+		}
+
+		for _, certificate := range certificateListForNamespace.Items {
+			if certificate.Status.Conditions[len(certificate.Status.Conditions)-1].Status == "True" && certificate.Status.Conditions[len(certificate.Status.Conditions)-1].Type == "Ready" && certificate.Status.Conditions[len(certificate.Status.Conditions)-1].Message == "Certificate is up to date and has not expired" {
```
Comment (davisgomes) [inline tools/vz/pkg/analysis/internal/util/cluster/certificates.go:49]: You could potentially simplify here by assigning this value to a variable:
```
latestCondition := certificate.Status.Conditions[len(certificate.Status.Conditions)-1]
if latestCondition.Status == "True" && ...
```

Comment (davisgomes) [inline tools/vz/pkg/analysis/internal/util/cluster/certificates.go:49]: Are status condition lists organized by time? I believe if a condition exists, it is overwritten. It is possible that the last condition is not the most recent, so you might want to loop through and verify using the timestamps.

File: tools/vz/pkg/analysis/internal/util/cluster/certificates.go [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+	}
+	defer file.Close()
+	fileBytes, err := io.ReadAll(file)
+	if err != nil {
+		log.Debugf("Failed reading Json file %s", path)
+		return nil, err
+	}
+	err = encjson.Unmarshal(fileBytes, &certList)
+	if err != nil {
+		log.Debugf("Failed to unmarshal CertificateList at %s", path)
```
Comment (davisgomes) [inline tools/vz/pkg/analysis/internal/util/cluster/certificates.go:84]: Same comment as above. Use `log.ErrorfNewError`

File: tools/vz/pkg/analysis/internal/util/cluster/certificates.go [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+// This function determines if the VPO is currently hanging due to certificate issues
+// It does this by checking the last 10 logs of the VPO and determines all the certificates that the VPO is hanging along
+// It returns a map containing these certificates as keys and their respective namespaces as values
+// This map is used by the main certificate analysis function to determine if the VPO is hanging on a valid certificate
+func determineIfVPOIsHangingDueToCerts(log *zap.SugaredLogger, clusterRoot string) (map[string]string, error) {
+	listOfCertificatesThatVPOIsHangingOn := make(map[string]string)
+	vpologRegExp := regexp.MustCompile(`verrazzano-install/verrazzano-platform-operator-.*/logs.txt`)
+	allPodFiles, err := files.GetMatchingFiles(log, clusterRoot, vpologRegExp)
+	if err != nil {
+		return listOfCertificatesThatVPOIsHangingOn, err
```
Comment (davisgomes) [inline tools/vz/pkg/analysis/internal/util/cluster/certificates.go:116]: Make sure to log every error that occurs from outside packages, or calls to functions that do not have access to the logger

File: tools/vz/pkg/analysis/internal/util/cluster/certificates.go [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+	if len(allPodFiles) == 0 {
+		return listOfCertificatesThatVPOIsHangingOn, nil
+	}
+	vpoLog := allPodFiles[0]
+	allMessages, err := files.ConvertToLogMessage(vpoLog)
+	if err != nil {
+		return listOfCertificatesThatVPOIsHangingOn, err
+	}
+	//If the VPO has greater than 10 messages, the last 10 logs are the input. Else, the whole VPO logs are the input
+	lastTenVPOLogs := allMessages[:]
```
Comment (davisgomes) [inline tools/vz/pkg/analysis/internal/util/cluster/certificates.go:127]: two comments:
Are you confident 10 is a large enough range to diagnose this issue?

This is nitpicky, but you could consider reducing complexity here using "min"
```
lastTenVPOLogs := allMessages[math.Min(0, len(allMessages) - 10):]
```

File: tools/vz/pkg/analysis/internal/util/cluster/certificates.go [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+	if len(allMessages) > 10 {
+		lastTenVPOLogs = allMessages[len(allMessages)-10:]
+	}
+	for _, VPOLog := range lastTenVPOLogs {
+		VPOLogMessage := VPOLog.Message
+		if strings.Contains(VPOLogMessage, "message: Issuing certificate as Secret does not exist") && strings.HasPrefix(VPOLogMessage, "Certificate ") {
+			VPOLogCertificateNameAndNamespace := strings.Split(VPOLogMessage, " ")[1]
+			namespaceAndCertificateNameSplit := strings.Split(VPOLogCertificateNameAndNamespace, "/")
+			nameSpace := namespaceAndCertificateNameSplit[0]
+			certificateName := namespaceAndCertificateNameSplit[1]
```
Comment (davisgomes) [inline tools/vz/pkg/analysis/internal/util/cluster/certificates.go:137]: This seems a bit shaky especially if we happen to update the string for how we report certificate errors. Are there any other ways to glean the certificate information from the logs. If not, I think this is doable for now.

File: tools/vz/pkg/analysis/internal/util/cluster/certificates.go (hunk 3 above)
Comment (George-Aeillo) [reply tools/vz/pkg/analysis/internal/util/cluster/certificates.go:49]: Hello! According to the cert-manager API, timestamps are optional with regard to conditions. What do you mean by if a condition exists, then it is overwritten?

File: tools/vz/pkg/analysis/internal/util/cluster/certificates.go [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+	certList := &certv1.CertificateList{}
+	file, err := os.Open(path)
+	if err != nil {
+		log.Debugf("file %s not found", path)
+		return nil, nil
+	}
+	defer file.Close()
+	fileBytes, err := io.ReadAll(file)
+	if err != nil {
+		log.Debugf("Failed reading Json file %s", path)
```
[the remark this answers is not in the document]
Comment (George-Aeillo) [reply tools/vz/pkg/analysis/internal/util/cluster/certificates.go:79]: Hello! log.ErrorfNewError does not appear to a function used by the suggaredLogger. Should I use log.Error instead?

File: tools/vz/pkg/analysis/internal/util/cluster/certificates.go (hunk 3 above)
Comment (George-Aeillo) [reply tools/vz/pkg/analysis/internal/util/cluster/certificates.go:49]: The conditions by themselves are optional for a certificate as well. Do you think I am able to assume that a certificate has a list of conditions, but the conditions have timestamps? This is how the certificates appear in a VZ cluster, so I assume this is default behavior. If this is the case, I will loop through and verify using the timestamps.

Comment (davisgomes) [reply tools/vz/pkg/analysis/internal/util/cluster/certificates.go:49]: Yea I think that should be the default. I guess you can ignore ones without timestamps.

File: tools/vz/pkg/analysis/internal/util/cluster/certificates.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (davisgomes) [reply tools/vz/pkg/analysis/internal/util/cluster/certificates.go:79]: Yes that should suffice. `log.ErrorfNewError` is used by the Verrazzano logger, not the suggaredLogger

[the remark this answers is not in the document]
Comment (George-Aeillo) [reply tools/vz/pkg/analysis/internal/util/cluster/certificates.go:79]: Thanks!

File: tools/vz/pkg/analysis/internal/util/cluster/certificates.go (hunk 3 above)
Comment (George-Aeillo) [reply tools/vz/pkg/analysis/internal/util/cluster/certificates.go:49]: Thanks!

File: tools/vz/pkg/analysis/internal/util/cluster/certificates.go (hunk 5 above)
Comment (George-Aeillo) [reply tools/vz/pkg/analysis/internal/util/cluster/certificates.go:116]: Thanks! In this example you highlighted the correct way of doing it, as files.GetMatchingFiles takes in the logger?

File: tools/vz/pkg/analysis/internal/util/cluster/certificates.go (hunk 6 above)
Comment (George-Aeillo) [reply tools/vz/pkg/analysis/internal/util/cluster/certificates.go:127]: In this issue, we are trying to detect if the VPO is currently hanging and waiting for certificates. In the pipeline failure that exposed the issue, the VPO was waiting for two certificates, but it should have been ready. 10 was an arbitrary number that was decided in a meeting. If the VPO is hanging on this issue, it appears to output the waiting for certificate message continuously. If 10 is not a large enough range, what range do you suggest? I have attached the link to the VPO Log file. 
https://build.verrazzano.io/job/verrazzano-new-oci-dns-acceptance-tests/job/release-1.4/521/artifact/oke-acceptance-tests-cluster-snapshot/full-cluster/cluster-snapshot/verrazzano-install/verrazzano-platform-operator-b6f6b55c6-7hdtr/logs.txt/*view*/

File: tools/vz/pkg/analysis/internal/util/cluster/certificates.go (hunk 5 above)
Comment (davisgomes) [reply tools/vz/pkg/analysis/internal/util/cluster/certificates.go:116]: Yes. That's correct

File: tools/vz/pkg/analysis/internal/util/cluster/certificates.go (hunk 7 above)
Comment (George-Aeillo) [reply tools/vz/pkg/analysis/internal/util/cluster/certificates.go:137]: Hello! From the logs, I can not think of another way to glean certificate information relating to this specific log messages. I have attached the VPO log file in a prior comment. Do you have any ideas how I could get this information in a less shaky way, if possible?

File: tools/vz/pkg/analysis/internal/util/cluster/certificates_test.go [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/verrazzano/verrazzano/tools/vz/pkg/analysis/internal/util/log"
+)
+
+// This tests asserts that no errors are called when the Certificate Analysis function is called on a valid input
+func TestAnalyzeCertificateIssues(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (George-Aeillo) [reply tools/vz/pkg/analysis/internal/util/cluster/certificates_test.go:13]: Hello! I was trying to follow the pattern that seemed to be in the files. This pattern seemed to be that for the resource_test.go files only tested if the function did not return an error for a valid input, while the main_test.go function seemed to test whether specific reports were generated for specific inputs. While I have added a specific case for a specific input in main_test.go, I plan on adding more specific test cases. Should I take these specific test cases out of main_test.go and add them to certificates_test.go. If not, what additional functionality should I add in certificates_test.go?

File: tools/vz/pkg/analysis/internal/util/cluster/certificates.go (hunk 5 above)
Comment (George-Aeillo) [reply tools/vz/pkg/analysis/internal/util/cluster/certificates.go:116]: Thanks!

File: tools/vz/pkg/analysis/internal/util/cluster/certificates.go [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+	err = encjson.Unmarshal(fileBytes, &certList)
+	if err != nil {
+		log.Debugf("Failed to unmarshal CertificateList at %s", path)
+		return nil, err
+	}
+	return certList, err
+}
+
+// This function creates the message and formally adds the issue to the IssueReporter
+func reportCertificateIssue(log *zap.SugaredLogger, clusterRoot string, certificate certv1.Certificate, issueReporter *report.IssueReporter, certificateFile string, VPOHangingIssue bool, isCertificateExpired bool) {
```
[the remark this answers is not in the document]
Comment (George-Aeillo) [reply tools/vz/pkg/analysis/internal/util/cluster/certificates.go:91]: Thanks!

File: tools/vz/pkg/analysis/internal/util/cluster/certificates.go [hunk 11]
```diff
@@ -0,0 +1,146 @@
...
+	files := []string{certificateFile}
+	if VPOHangingIssue {
+		message := []string{"The VPO is hanging due to a long time for the certificate to complete, but the certificate named " + certificate.ObjectMeta.Name + " in namespace " + certificate.ObjectMeta.Namespace + " is ready"}
```
[the remark this answers is not in the document]
Comment (George-Aeillo) [reply tools/vz/pkg/analysis/internal/util/cluster/certificates.go:94]: Thanks!

File: tools/vz/pkg/analysis/internal/util/cluster/certificates.go (hunk 6 above)
Comment (davisgomes) [reply tools/vz/pkg/analysis/internal/util/cluster/certificates.go:127]: I think that's okay as long as you've aired it out. I just want to make sure that was verified, but it seems okay to me.

File: tools/vz/pkg/analysis/internal/util/cluster/certificates.go (hunk 7 above)
Comment (davisgomes) [reply tools/vz/pkg/analysis/internal/util/cluster/certificates.go:137]: No I think you're right. This seems to be the only sure way to get the name and namespace. I think you are fine to leave it as is.

File: tools/vz/pkg/analysis/internal/util/cluster/certificates_test.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (davisgomes) [reply tools/vz/pkg/analysis/internal/util/cluster/certificates_test.go:13]: The `*_test.go` file should only test functions that are in the corresponding `*.go` file. Generally, there is one unit test function for each function in the Go file. You don't have to create a test for each function if its functionality is generally static, and most lines are tested from other unit tests. Just try to get as many lines covered by unit tests as you can, especially edge cases.

File: tools/vz/pkg/analysis/internal/util/cluster/certificates.go (hunk 6 above)
Comment (George-Aeillo) [reply tools/vz/pkg/analysis/internal/util/cluster/certificates.go:127]: Thanks!

File: tools/vz/pkg/analysis/internal/util/cluster/certificates.go (hunk 7 above)
Comment (George-Aeillo) [reply tools/vz/pkg/analysis/internal/util/cluster/certificates.go:137]: Thanks!

File: tools/vz/pkg/helpers/vzcapture.go [hunk 12]
```diff
@@ -284,6 +289,22 @@ func captureWorkLoads(kubeClient kubernetes.Interface, namespace, captureDir str
 	return nil
 }
 
+// This function gets the certificates from the client for the current namespace and outputs the objects to a certificates.json file, if certificates are present in that namespace.
+func captureCertificates(client clipkg.Client, namespace, captureDir string, vzHelper VZHelper) error {
+	certificateList := v1.CertificateList{}
```
[the remark this answers is not in the document]
Comment (George-Aeillo) [reply tools/vz/pkg/helpers/vzcapture.go:294]: Yes, this function just captures the cert object and outputs them to a certificates.json file and sanitizes them for ip address and hostname. Through this change, all resources generated by the bug report will be sanitized for IP and for hostname.

File: tools/vz/pkg/analysis/internal/util/cluster/certificates.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (George-Aeillo) [reply tools/vz/pkg/analysis/internal/util/cluster/certificates.go]: Thanks! I have made this change.

File: tools/vz/pkg/analysis/internal/util/cluster/certificates.go [hunk 13]
```diff
@@ -0,0 +1,146 @@
...
+		return listOfCertificatesThatVPOIsHangingOn, nil
+	}
+	vpoLog := allPodFiles[0]
+	allMessages, err := files.ConvertToLogMessage(vpoLog)
+	if err != nil {
+		return listOfCertificatesThatVPOIsHangingOn, err
+	}
+	//If the VPO has greater than 10 messages, the last 10 logs are the input. Else, the whole VPO logs are the input
+	lastTenVPOLogs := allMessages[:]
+	if len(allMessages) > 10 {
```
[the remark this answers is not in the document]
Comment (George-Aeillo) [reply tools/vz/pkg/analysis/internal/util/cluster/certificates.go]: This function doesn't look at the last ten messages for a particular certificate, it looks at the ten VPO logs for the entire VPO. So if the certificate is currently in a pending state when the capture is taken, it would report that the cert is currently not ready regardless of what the VPO log says because the status of the certificate is not ready. The VPO log logic only kicks in if certificates are in a ready state, but the VPO is hanging on the already ready certificates.

File: tools/vz/pkg/helpers/vzcapture_test.go [hunk 14]
```diff
@@ -383,3 +389,26 @@ func TestGetPodListAll(t *testing.T) {
 	assert.NoError(t, err)
 	assert.Equal(t, podLength, len(pods))
 }
+
+//	 	GIVEN a k8s cluster with certificates present in a namespace  ,
+//		WHEN I call functions to create a list of certificates for the namespace,
+//		THEN expect it to write to the provided resource file and no error should be returned.
+func TestCreateCertificateFile(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (George-Aeillo) [reply tools/vz/pkg/helpers/vzcapture_test.go:396]: Thanks!

File: tools/vz/pkg/analysis/internal/util/cluster/certificates.go [hunk 15]
```diff
@@ -0,0 +1,146 @@
...
+	}
+	message := []string{"The certificate named " + certificate.ObjectMeta.Name + " in namespace " + certificate.ObjectMeta.Namespace + " is not valid and experiencing issues"}
+	issueReporter.AddKnownIssueMessagesFiles(report.CertificateExperiencingIssuesInCluster, clusterRoot, message, files)
+}
+
+// This function determines if the VPO is currently hanging due to certificate issues
+// It does this by checking the last 10 logs of the VPO and determines all the certificates that the VPO is hanging along
+// It returns a map containing these certificates as keys and their respective namespaces as values
+// This map is used by the main certificate analysis function to determine if the VPO is hanging on a valid certificate
+func determineIfVPOIsHangingDueToCerts(log *zap.SugaredLogger, clusterRoot string) (map[string]string, error) {
```
[the remark this answers is not in the document]
Comment (George-Aeillo) [reply tools/vz/pkg/analysis/internal/util/cluster/certificates.go:115]: By report, do you mean analysis report or bug report? The bug report currently captures the status. This currently checks to see if the VPO is hanging on certificates and the results of this function only kick in if a certificate is in a ready status. If the certificate is not in a ready status, the analysis report will report that certificate as having issues. The VPO logic is to solve the issue of if certificates are ready, but the VPO is hanging due to the long time it took for the certificate to get approved. Is this the intended behavior? How would I go about linking this to certificates that are not in a ready status?

[the remark this answers is not in the document]
Comment (michael-cico) [reply tools/vz/pkg/analysis/internal/util/cluster/certificates.go:115]: I might be misunderstanding it a bit... I'm not sure how we'd get in a state where the VPO where component code is waiting on certs that are ready, but then again it's probably worth noting in the analysis then.  In my head the VPO is  stuck reconciling checking for certificates that are still pending (not Ready).

The VPO never gets "hung" per se, but keeps re-queuing/reconciling and calling component `IsReady()` checks.  It will never finish reconciling while the certs are not ready, but once they are it should finish.

I might be splitting hairs here... @davisgomes does this make sense to you, or am I missing the mark?

[the remark this answers is not in the document]
Comment (George-Aeillo) [reply tools/vz/pkg/analysis/internal/util/cluster/certificates.go:115]: I think the issue has to do with the VPO timing out, so it will continually report a certificate as not ready, even if the certificate may be ready. This has occurred before, the test files that I have made as part of this commit came from a situation where this happened.

[the remark this answers is not in the document]
Comment (michael-cico) [reply tools/vz/pkg/analysis/internal/util/cluster/certificates.go:115]: OK, I thought that the case was the certs weren't ready (like with Let's Encrypt staging, or prod where we exceed the rate limit, etc), and that was causing the reconcile to never finish.  

Technically the VPO never times out, but the CLI operation does.  The VPO keeps trying to reconcile the CR in the cluster.

File: tools/vz/pkg/analysis/internal/util/cluster/certificates.go (hunk 2 above)
Comment (George-Aeillo) [reply tools/vz/pkg/analysis/internal/util/cluster/certificates.go]: Thanks!

File: tools/vz/pkg/analysis/internal/util/cluster/certificates.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (George-Aeillo) [reply tools/vz/pkg/analysis/internal/util/cluster/certificates.go:115]: Thanks! I have a case from the pipeline that has this hanging issue occurring and it is currently incorporated in my unit tests.

**Review by davisgomes — DISMISSED** [review]

File: tools/vz/pkg/helpers/vzhelper.go [hunk 16]
```diff
@@ -210,6 +212,7 @@ func NewScheme() *runtime.Scheme {
 	_ = networkingv1.AddToScheme(scheme)
 	_ = oam.AddToScheme(scheme)
 	_ = batchv1.AddToScheme(scheme)
+	_ = certv1.AddToScheme(scheme)
```
[the remark this answers is not in the document]
Comment (George-Aeillo) [reply tools/vz/pkg/helpers/vzhelper.go:215]: This change is for allowing the controller runtime client to access cert-manager resources, as certificates in cert-manager are custom resources and are not accessible to a standard kubeclient.

File: tools/vz/pkg/analysis/internal/util/cluster/analyzer.go [hunk 17]
```diff
@@ -30,9 +30,12 @@ import (
 //	So we may break this into 2 lists in the future: serial analysis functions, parallel analysis functions
 //	Analyzers that may fall into this category should be annotated, with a comment, there currently is only
 //	one that may require that.
+
+// These are the high level analysis functions that are called. The "Runtime Issues" maps to only certificate functions currently.
 var clusterAnalysisFunctions = map[string]func(log *zap.SugaredLogger, directory string) (err error){
 	"Verrazzano Status":  AnalyzeVerrazzano, // Execute first, this may share data other analyzers can use
 	"Pod Related Issues": AnalyzePodIssues,
+	"Runtime Issues":     AnalyzeCertificateRelatedIssues,
```
[the remark this answers is not in the document]
Comment (George-Aeillo) [reply tools/vz/pkg/analysis/internal/util/cluster/analyzer.go:38]: Hello! I was suggested to name it "Runtime Issues", as other functions that are not related to certificates may be called in the future by this top level function, but only certificates issues are being used in it currently. Should I still rename it as you suggested, or should I keep its current name of "Runtime Issues"?

File: tools/vz/pkg/analysis/internal/util/cluster/certificates.go [hunk 18]
```diff
@@ -0,0 +1,199 @@
...
+	}
+	allNamespacesFound, err = files.FindNamespaces(log, clusterRoot)
+	if err != nil {
+		return err
+	}
+	var issueReporter = report.IssueReporter{
+		PendingIssues: make(map[string]report.Issue),
+	}
+	for _, namespace := range allNamespacesFound {
+		certificateFile := files.FindFileInNamespace(clusterRoot, namespace, "certificates.json")
```
[the remark this answers is not in the document]
Comment (George-Aeillo) [reply tools/vz/pkg/analysis/internal/util/cluster/certificates.go]: I have altered the code to use constants, as appropriate.

File: tools/vz/pkg/analysis/internal/util/cluster/certificates.go [hunk 19]
```diff
@@ -0,0 +1,199 @@
...
+	"strings"
+	"time"
+
+	certv1 "github.com/cert-manager/cert-manager/pkg/apis/certmanager/v1"
+	"github.com/verrazzano/verrazzano/tools/vz/pkg/analysis/internal/util/files"
+	"github.com/verrazzano/verrazzano/tools/vz/pkg/analysis/internal/util/report"
+	"go.uber.org/zap"
+)
+
+// This is the initial entry function for certificate related issues.
```
[the remark this answers is not in the document]
Comment (George-Aeillo) [reply tools/vz/pkg/analysis/internal/util/cluster/certificates.go]: I have altered my function conventions to follow that pattern, as the comments now not only explain what the function does, but they also have the function name as part of the comment.

File: tools/vz/pkg/analysis/test/cluster/testCLIHangingIssue/cluster-snapshot/verrazzano-install/verrazzano-platform-operator-b6f6b55c6-7hdtr/logs.txt [hunk 20]
```diff
@@ -0,0 +1,2014 @@
...
+{"level":"info","@timestamp":"2023-03-08T14:29:21.220Z","caller":"verrazzano/install.go:139","message":"Component verrazzano-application-operator install started ","resource_namespace":"default","resource_name":"my-verrazzano","controller":"verrazzano","component":"verrazzano-application-operator","operation":"install"}
+{"level":"info","@timestamp":"2023-03-08T14:29:21.221Z","caller":"helm/helmcli.go:227","message":"Running Helm command /usr/bin/helm upgrade verrazzano-application-operator /verrazzano/platform-operator/helm_config/charts/verrazzano-application-operator --namespace verrazzano-system --install -f /tmp/helm-overrides-verrazzano-2371951217.yaml for release verrazzano-application-operator","resource_namespace":"default","resource_name":"my-verrazzano","controller":"verrazzano","component":"verrazzano-application-operator","operation":"install"}
+2023/03/08 14:29:23 http: TLS handshake error from 10.244.0.0:49224: EOF
```
Comment (bhatpmk) [inline tools/vz/pkg/analysis/test/cluster/testCLIHangingIssue/cluster-snapshot/verrazzano-install/verrazzano-platform-operator-b6f6b55c6-7hdtr/logs.txt:192]: In general, mast IP address from the test data

File: tools/vz/pkg/analysis/internal/util/report/issue.go [hunk 21]
```diff
@@ -149,6 +152,9 @@ var knownIssues = map[string]Issue{
 	ExternalDNSConfigureIssue:    {Type: ExternalDNSConfigureIssue, Summary: "Failed to setup DNS configuration", Informational: false, Impact: 10, Confidence: 10, Actions: []Action{KnownActions[ExternalDNSConfigureIssue]}},
 	KeycloakDataMigrationFailure: {Type: KeycloakDataMigrationFailure, Summary: "Failure(s) migrating Keycloak data during MySQL upgrade", Informational: true, Impact: 10, Confidence: 10, Actions: []Action{KnownActions[KeycloakDataMigrationFailure]}},
 	KontainerDriverNotReady:      {Type: KontainerDriverNotReady, Summary: "Rancher KontainerDriver resources are not in the expected state", Informational: false, Impact: 10, Confidence: 10, Actions: []Action{KnownActions[KontainerDriverNotReady]}},
+	VZClientHangingIssueDueToLongCertificateApproval: {Type: VZClientHangingIssueDueToLongCertificateApproval, Summary: " VZ Client is hanging due to the long time that it takes to approve and provision certificates", Informational: true, Impact: 10, Confidence: 10},
```
Comment (bhatpmk) [inline tools/vz/pkg/analysis/internal/util/report/issue.go:155]: Replace VZ with Verrazzano in the summary.

File: tools/vz/pkg/analysis/test/cluster/testCLIHangingIssue/cluster-snapshot/verrazzano-install/verrazzano-platform-operator-b6f6b55c6-7hdtr/logs.txt (hunk 20 above)
Comment (George-Aeillo) [reply tools/vz/pkg/analysis/test/cluster/testCLIHangingIssue/cluster-snapshot/verrazzano-install/verrazzano-platform-operator-b6f6b55c6-7hdtr/logs.txt:192]: Thanks!

File: tools/vz/pkg/analysis/internal/util/cluster/certificates.go [hunk 22]
```diff
@@ -0,0 +1,198 @@
...
+	certList := &certv1.CertificateList{}
+	file, err := os.Open(path)
+	if err != nil {
+		log.Debug("file %s not found", path)
+		return nil, nil
+	}
+	defer file.Close()
+	fileBytes, err := io.ReadAll(file)
+	if err != nil {
+		log.Error("Failed reading Json file %s", path)
```
[the remark this answers is not in the document]
Comment (George-Aeillo) [reply tools/vz/pkg/analysis/internal/util/cluster/certificates.go:103]: How would I verify this error?

[the remark this answers is not in the document]
Comment (George-Aeillo) [reply tools/vz/pkg/analysis/internal/util/cluster/certificates.go]: This work relating to cli logging issues and consistency on how errors are logged will be treated as a separate bug.

**Review by davisgomes — APPROVED** [review]

**Review by davisgomes — APPROVED** [review]
